internal/test/builder/builders.go

```diff
@@ -990,3 +990,29 @@ func (m *MachineHealthCheckBuilder) Build() *clusterv1.MachineHealthCheck {
 		},
 	}
 }
+
+// Clone returns a copy of the MachineHealthCheckBuilder.
+func (m *MachineHealthCheckBuilder) Clone() *MachineHealthCheckBuilder {
+	out := &MachineHealthCheckBuilder{
+		name:        m.name,
+		namespace:   m.namespace,
+		clusterName: m.clusterName,
+		selector:    *m.selector.DeepCopy(),
+	}
+
+	if m.maxUnhealthy != nil {
+		maxUnhealthy := *m.maxUnhealthy
+		out.maxUnhealthy = &maxUnhealthy
+	}
+
+	if len(m.conditions) > 0 {
+		out.conditions = make([]clusterv1.UnhealthyCondition, len(m.conditions))
+		copy(out.conditions, m.conditions)
+	}
+	if len(m.ownerRefs) > 0 {
+		for _, ref := range m.ownerRefs {
+			out.ownerRefs = append(out.ownerRefs, *ref.DeepCopy())
+		}
+	}
```

Comment:
nit: why do we copy those arrays differently?

Reply:
My understanding is that because OwnerRefs have pointers in them so they need to be deepcopied to ensure we're getting a copy. UnhealthyConditions only contains values so it should be alright.

Reply:
Thx. I'll take a look but that sounds reasonable :)

Reply:
The alternative here (and maybe a better option) would be to add deepcopy generation for this file.

Reply:
Yup, in the bets case it just generates everything for us and if we're fine with calling DeepCopy instead of Clone we don't have to implement anything ourselves/manually.

WDYT, worth a try?

Reply:
It seems the kubebuilder generation doesn't work with map[string]interface{}.

Reply:
I disabled generation for those fields - I'm not sure if there's a way to tweak it to make generation of map[string]interface{} work for kubebuilder or if it would be better to stop using that for the spec fields in the builders. For now though those functions aren't needed so I think it's fine to go forward with the PR as it stands.

Reply:
Sounds great so far and looks good at a first glance (will take a closer look tomorrow).

I assume "kubebuilder generation" = the deep copy generation of controller-gen? (just to confirm)

Reply:
Yeah - I'm doing it here using the kubebuilder tags.

Reply:
I think we won't be able to make that work (at least without adjusting controller-gen or changing the data type of the values of the map).

The problem is that controller-gen cannot generate a DeepCopy func for a `interface{}` (I guess because it can be anything):
https://github.com/kubernetes-sigs/controller-tools/blob/a410b23ff25739563e103e98af887adfd05feb98/pkg/deepcopy/traverse.go#L386-L393